### PR TITLE
mds: trim mdlog when segments exceed threshold and trim was idle

### PR DESCRIPTION
**Pull request by vshankar**
Should have explained the intent in the commit message, but pushing this out for reviews. See https://tracker.ceph.com/issues/66948#note-9 for an explanation.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by vshankar** [comment]
@gregsfortytwo - from your question in tracker

> This is bizarre, why is the mdlog leaving expired_segments just sitting in the log? Is that formula just wrong, or being used for two purposes that don't quite mesh or something?

I have to check that but it has to do something with minor segment boundary which I haven't looked at closely. This change might not be required if there is a bug in MDLog::_trim_expired_segments() or if it's keeping segments around for more time than expected.

**Comment by vshankar** [comment]
From /a/vshankar-2024-07-08_07:21:13-fs-wip-vshankar-testing-20240705.150505-debug-testing-default-smithi/7791866 where the trim warning was seen, for ./remote/smithi089/log/ceph-mds.f.log.1.gz, there is a flurry of the below log messages for sometime

```
2024-07-09T23:43:48.946+0000 7f2ed139b640 20 mds.2.log _trim_expired_segments: examining LogSegment(258092/0x1e72da941 events=5)
2024-07-09T23:43:48.946+0000 7f2ed139b640 10 mds.2.log _trim_expired_segments: maybe expiring LogSegment(258092/0x1e72da941 events=5)
2024-07-09T23:43:48.946+0000 7f2ed139b640 20 mds.2.log _trim_expired_segments: examining LogSegment(258097/0x1e789b9ba events=111)
2024-07-09T23:43:48.946+0000 7f2ed139b640 10 mds.2.log _trim_expired_segments: maybe expiring LogSegment(258097/0x1e789b9ba events=111)
2024-07-09T23:43:48.946+0000 7f2ed139b640 20 mds.2.log _trim_expired_segments: examining LogSegment(258208/0x1e7c4152e events=39)
2024-07-09T23:43:48.946+0000 7f2ed139b640 10 mds.2.log _trim_expired_segments: maybe expiring LogSegment(258208/0x1e7c4152e events=39)
2024-07-09T23:43:48.946+0000 7f2ed139b640 20 mds.2.log _trim_expired_segments: examining LogSegment(258247/0x1e8013480 events=39)
2024-07-09T23:43:48.946+0000 7f2ed139b640 10 mds.2.log _trim_expired_segments: maybe expiring LogSegment(258247/0x1e8013480 events=39)
...
...
...
2024-07-09T23:43:48.946+0000 7f2ed139b640 10 mds.2.log _trim_expired_segments: maybe expiring LogSegment(267055/0x21583f596 events=126)
2024-07-09T23:43:48.946+0000 7f2ed139b640 20 mds.2.log _trim_expired_segments: examining LogSegment(267181/0x215c1638c events=60)
2024-07-09T23:43:48.946+0000 7f2ed139b640 10 mds.2.log _trim_expired_segments: maybe expiring LogSegment(267181/0x215c1638c events=60)
2024-07-09T23:43:48.946+0000 7f2ed139b640 20 mds.2.log _trim_expired_segments: examining LogSegment(267241/0x21602de98 events=42)
2024-07-09T23:43:48.946+0000 7f2ed139b640 10 mds.2.log _trim_expired_segments: maybe expiring LogSegment(267241/0x21602de98 events=42)
2024-07-09T23:43:48.946+0000 7f2ed139b640 20 mds.2.log _trim_expired_segments: examining LogSegment(267283/0x216df1773 events=54)
2024-07-09T23:43:48.946+0000 7f2ed139b640 10 mds.2.log _trim_expired_segments waiting for expiry LogSegment(267283/0x216df1773 events=54)
```

_trim_expired_segments() removes logseg's from expired_segments only when it sees a major segment. Its been a while that I've looked at the minor/major segment parts in mdlog (last involvement was when the PR was under review) -- I'll have to revisit it.

**Comment by batrick** [comment]
```
2024-07-09T23:36:06.567+0000 7f2ed5ba4640 20 mds.2.log _submit_entry EExport 0x100000022e5 to mds.0 [metablob 0x1, 8 dirs]
2024-07-09T23:36:06.567+0000 7f2ed5ba4640 10 mds.2.log _segment_upkeep: starting new segment, current LogSegment(267181/0x215c1638c events=60)
2024-07-09T23:36:06.567+0000 7f2ed5ba4640 20 mds.2.log _submit_entry ESegment(0)
```

The actual problem is that these minor segments are consistently small by event count. This is caused by the large `EExport` events. That's causing new segments to be created via:

https://github.com/ceph/ceph/blob/2966f225ee943a98cb5a4d7a6c429d728747c864/src/mds/MDLog.cc#L410

I think the code needs to be not care about the _events_ since the last major segment but instead the number of minor segments since the last major segment:

https://github.com/ceph/ceph/blob/2966f225ee943a98cb5a4d7a6c429d728747c864/src/mds/MDLog.cc#L406

So if there has been ~8-16 segments since the last major segment, write a new one.

**Comment by batrick** [comment]
Note this is only found because we have export thrashing turned on.

**Review by gregsfortytwo — DISMISSED** [review]
I don't like this solution because it's basically introducing a post-hoc patch-up, and it's not synchronized with what the Beacon health check is doing, so it could still flash up a warning and then clean up. It's much better if we can follow some straightforward length rules that are checked for in the same way in all relevant parts of the code.


Does Patrick's formula resolve this on its own, or is more needed?

**Comment by vshankar** [comment]
> > Note this is only found because we have export thrashing turned on.
> 
> Downstream QE is able to reproduce this and I don't think their tests were thrashing MDSs.

EDIT: thrashing exports.

**Comment by vshankar** [comment]
> ```
> 2024-07-09T23:36:06.567+0000 7f2ed5ba4640 20 mds.2.log _submit_entry EExport 0x100000022e5 to mds.0 [metablob 0x1, 8 dirs]
> 2024-07-09T23:36:06.567+0000 7f2ed5ba4640 10 mds.2.log _segment_upkeep: starting new segment, current LogSegment(267181/0x215c1638c events=60)
> 2024-07-09T23:36:06.567+0000 7f2ed5ba4640 20 mds.2.log _submit_entry ESegment(0)
> ```
> 
> The actual problem is that these minor segments are consistently small by event count but too large due to the large `EExport` events. That's causing new segments to be created via:
> 
> https://github.com/ceph/ceph/blob/2966f225ee943a98cb5a4d7a6c429d728747c864/src/mds/MDLog.cc#L410
> 
> I think the code needs to not care about the _events_ since the last major segment but instead the number of minor segments since the last major segment:
> 
> https://github.com/ceph/ceph/blob/2966f225ee943a98cb5a4d7a6c429d728747c864/src/mds/MDLog.cc#L406
> 
> So if there has been ~8-16 segments since the last major segment, write a new one.

This would create major segments (`ESubtreeMap`) more frequently. It's definitely better than what it was originally (which was to log an `ESubtreeMap` event) when:

```
  } else if (ls->end/period != ls->offset/period ||
             ls->num_events >= g_conf()->mds_log_events_per_segment) {
    dout(10) << "submit_entry also starting new segment: last = "
             << ls->seq  << "/" << ls->offset << ", event seq = " << event_seq << dendl;
    _start_new_segment();
```

Now, logging a major segment boundary after 8 (or whatever) number of minor segments would avoid long wait periods when waiting for segments to expire to eventually trim them, but its still looks like a relatively large increase from logging a major segment after (1024 * 12) events. And the downstream reproducer didn't involve export thrashing (I will have to double check), so I'm still curious if this is the only thing we need to workaround the trim issue.

**Comment by batrick** [comment]
> > ```
> > 2024-07-09T23:36:06.567+0000 7f2ed5ba4640 20 mds.2.log _submit_entry EExport 0x100000022e5 to mds.0 [metablob 0x1, 8 dirs]
> > 2024-07-09T23:36:06.567+0000 7f2ed5ba4640 10 mds.2.log _segment_upkeep: starting new segment, current LogSegment(267181/0x215c1638c events=60)
> > 2024-07-09T23:36:06.567+0000 7f2ed5ba4640 20 mds.2.log _submit_entry ESegment(0)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The actual problem is that these minor segments are consistently small by event count but too large due to the large `EExport` events. That's causing new segments to be created via:
> > https://github.com/ceph/ceph/blob/2966f225ee943a98cb5a4d7a6c429d728747c864/src/mds/MDLog.cc#L410
> > 
> > I think the code needs to not care about the _events_ since the last major segment but instead the number of minor segments since the last major segment:
> > https://github.com/ceph/ceph/blob/2966f225ee943a98cb5a4d7a6c429d728747c864/src/mds/MDLog.cc#L406
> > 
> > So if there has been ~8-16 segments since the last major segment, write a new one.
> 
> This would create major segments (`ESubtreeMap`) more frequently. It's definitely better than what it was originally (which was to log an `ESubtreeMap` event) when:
> 
> ```
>   } else if (ls->end/period != ls->offset/period ||
>              ls->num_events >= g_conf()->mds_log_events_per_segment) {
>     dout(10) << "submit_entry also starting new segment: last = "
>              << ls->seq  << "/" << ls->offset << ", event seq = " << event_seq << dendl;
>     _start_new_segment();
> ```
> 
> Now, logging a major segment boundary after 8 (or whatever) number of minor segments would avoid long wait periods when waiting for segments to expire to eventually trim them, but its still looks like a relatively large increase from logging a major segment after (1024 * 12) events. And the downstream reproducer didn't involve export thrashing (I will have to double check), so I'm still curious if this is the only thing we need to workaround the trim issue.

Perhaps the downstream issue is slightly different?

**Comment by vshankar** [comment]
@batrick - Regarding the downstream issue related to trim, there are hints of directories being exported, but it should not be as wild as the export thrash test. The node where the logs are located is a bit unresponsive atm - will details the numbers when I have them.

File: src/mds/MDLog.cc [hunk 1]
```diff
@@ -736,12 +738,14 @@ void MDLog::trim()
       break;
     }
 
+    auto interval = std::chrono::duration<double>(last_trim - trim_start);
+    auto should_trim = is_oversegmented() && (interval.count() >= oversegmented_idle_interval);
```
Comment (mchangir) [inline src/mds/MDLog.cc:742]: Should the timestamp difference expression be `trim_start - last_trim` instead ?
I think the expression `last_trim - trim_start` might turn out negative.

**Comment by vshankar** [comment]
> @batrick - Regarding the downstream issue related to trim, there are hints of directories being exported, but it should not be as wild as the export thrash test. The node where the logs are located is a bit unresponsive atm - will details the numbers when I have them.

The downstream logs too have traces of directory export, so I'm intended to believe that what @batrick mentioned in https://github.com/ceph/ceph/pull/60381#issuecomment-2420181238 is what happening in the cluster.

I also want to discuss if the MDS really needs to expire upto a particular major segment? (before introducing major/minor segments, the MDS would just check if a segment is in expired_segment list before marking it a candidate for trimming (by setting the expire_pos in the journaler).

File: src/mds/MDLog.cc (hunk 1 above)
Comment (vshankar) [reply src/mds/MDLog.cc:742]: That should only be valid for the first iteration.

**Comment by batrick** [comment]
> > @batrick - Regarding the downstream issue related to trim, there are hints of directories being exported, but it should not be as wild as the export thrash test. The node where the logs are located is a bit unresponsive atm - will details the numbers when I have them.
> 
> The downstream logs too have traces of directory export, so I'm intended to believe that what @batrick mentioned in [#60381 (comment)](https://github.com/ceph/ceph/pull/60381#issuecomment-2420181238) is what happening in the cluster.
> 
> I also want to discuss if the MDS really needs to expire upto a particular major segment? (before introducing major/minor segments, the MDS would just check if a segment is in expired_segment list before marking it a candidate for trimming (by setting the expire_pos in the journaler).

The point of major/minor segments is to enforce the constraint that the MDS always begins replay with a major segment. We cannot trim minor segments except up to the next major segment.

**Comment by vshankar** [comment]
> > > @batrick - Regarding the downstream issue related to trim, there are hints of directories being exported, but it should not be as wild as the export thrash test. The node where the logs are located is a bit unresponsive atm - will details the numbers when I have them.
> > 
> > 
> > The downstream logs too have traces of directory export, so I'm intended to believe that what @batrick mentioned in [#60381 (comment)](https://github.com/ceph/ceph/pull/60381#issuecomment-2420181238) is what happening in the cluster.
> > I also want to discuss if the MDS really needs to expire upto a particular major segment? (before introducing major/minor segments, the MDS would just check if a segment is in expired_segment list before marking it a candidate for trimming (by setting the expire_pos in the journaler).
> 
> The point of major/minor segments is to enforce the constraint that the MDS always begins replay with a major segment. We cannot trim minor segments except up to the next major segment.

ACK. I was doubting that if we do `journaler->set_expire_pos()` for a segment and then crash, but the expire pos gets written, but I see a write_head() is done when wirtes out the updated positions. So we are good. I'll make the changes related to major segments.

Question: Any value in additionally keeping the current changes? (as a failsafe probably)

File: src/mds/MDLog.cc [hunk 2]
```diff
@@ -656,6 +659,10 @@ void MDLog::try_to_commit_open_file_table(uint64_t last_seq)
   }
 }
 
+bool MDLog::is_oversegmented() const {
```
Comment (batrick) [inline src/mds/MDLog.cc:662]:
```suggestion
bool MDLog::is_oversegmented() const {
```

"oversegmented" to me means something like "over fragmented"? I think just `::is_trimming_slow` would perhaps be better.

File: src/mds/Beacon.cc [hunk 3]
```diff
@@ -318,16 +318,15 @@ void Beacon::notify_health(MDSRank const *mds)
   // Detect MDS_HEALTH_TRIM condition
   // Indicates MDS is not trimming promptly
   {
-    const auto log_max_segments = mds->mdlog->get_max_segments();
-    const auto log_warn_factor = g_conf().get_val<double>("mds_log_warn_factor");
-    if (mds->mdlog->get_num_segments() > (size_t)(log_max_segments * log_warn_factor)) {
+    if (mds->mdlog->is_oversegmented()) {
```
Comment (batrick) [inline src/mds/Beacon.cc:321]: I like moving the logic to MDLog.

File: src/common/options/mds.yaml.in [hunk 4]
```diff
@@ -1741,3 +1741,13 @@ options:
   - mds
   flags:
   - runtime
+- name: mds_log_oversegmented_trim_idle_interval
```
Comment (batrick) [inline src/common/options/mds.yaml.in:1744]: this config and the related code I don't think we need?

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/common/options/mds.yaml.in (hunk 4 above)
Comment (vshankar) [reply src/common/options/mds.yaml.in]: Removed.

File: src/mds/MDLog.cc (hunk 2 above)
Comment (vshankar) [reply src/mds/MDLog.cc]: Changed.

File: src/common/options/mds.yaml.in [hunk 5]
```diff
@@ -586,16 +586,6 @@ options:
   min: 1
   services:
   - mds
-- name: mds_log_major_segment_event_ratio
```
Comment (batrick) [inline src/common/options/mds.yaml.in:589]: this change broke the docs

File: src/mds/MDLog.cc [hunk 6]
```diff
@@ -357,14 +358,13 @@ void MDLog::_submit_entry(LogEvent *le, MDSLogContextBase* c)
...
-  events_since_last_major_segment++;
 
   if (auto sb = dynamic_cast<SegmentBoundary*>(le); sb) {
     auto ls = _start_new_segment(sb);
     if (sb->is_major_segment_boundary()) {
       major_segments.insert(ls->seq);
       logger->set(l_mdl_segmjr, major_segments.size());
-      events_since_last_major_segment = 0;
+      minor_segments_since_last_major_segment = 0;
     }
```
Comment (batrick) [inline src/mds/MDLog.cc:368]:
```suggestion
    } else {
      ++minor_segments_since_last_major_segment;
    }
```

more appropriate place for this

File: src/mds/MDLog.cc [hunk 7]
```diff
@@ -403,14 +403,15 @@ void MDLog::_segment_upkeep()
...
-  if (events_since_last_major_segment > events_per_segment*major_segment_event_ratio) {
+  if (minor_segments_since_last_major_segment > minor_segments_per_major_segment) {
     dout(10) << __func__ << ": starting new major segment, current " << *ls << dendl;
     auto sle = mds->mdcache->create_subtree_map();
     _submit_entry(sle, NULL);
   } else if (ls->end/period != ls->offset/period || ls->num_events >= events_per_segment) {
     dout(10) << __func__ << ": starting new segment, current " << *ls << dendl;
     auto sb = new ESegment();
     _submit_entry(sb, nullptr);
+    ++minor_segments_since_last_major_segment;
```
Comment (batrick) [inline src/mds/MDLog.cc:414]:
```suggestion
```

File: src/mds/MDLog.cc [hunk 8]
```diff
@@ -1487,7 +1491,6 @@ void MDLog::_replay_thread()
       if (sb->is_major_segment_boundary()) {
         major_segments.insert(event_seq);
         logger->set(l_mdl_segmjr, major_segments.size());
-        events_since_last_major_segment = 0;
```
Comment (batrick) [inline src/mds/MDLog.cc:1490]: need to update `minor_segments_since_last_major_segment` here too, no?

**Review by batrick — CHANGES_REQUESTED** [review]
otherwise LGTM

File: src/mds/MDLog.cc (hunk 8 above)
Comment (vshankar) [reply src/mds/MDLog.cc:1490]: good catch, yes!

File: src/mds/MDLog.cc (hunk 6 above)
Comment (vshankar) [reply src/mds/MDLog.cc:368]: Agreed.

**Comment by vshankar** [comment]
jenkins test api

File: doc/cephfs/mds-journaling.rst [hunk 9]
```diff
@@ -141,15 +141,6 @@ The targetted size of a log segment in terms of number of events is controlled b
...
 .. confval:: mds_log_events_per_segment
 
-The frequency of major segments (noted by the journaling of the latest ``ESubtreeMap``) is controlled by:
-
-.. confval:: mds_log_major_segment_event_ratio
-
-When ``mds_log_events_per_segment * mds_log_major_segment_event_ratio``
-non-``ESubtreeMap`` events are logged, the MDS will journal a new
-``ESubtreeMap``. This is necessary to allow the journal to shrink in size
-during the trimming of expired segments.
```
Comment (batrick) [inline doc/cephfs/mds-journaling.rst:151]: Why not update this with the new `mds_log_minor_segments_per_major_segment`?

**Review by batrick — CHANGES_REQUESTED** [review]

File: doc/cephfs/mds-journaling.rst (hunk 9 above)
Comment (vshankar) [reply doc/cephfs/mds-journaling.rst:151]: Not sure why I didn't add that....

**Review by batrick — APPROVED** [review]

File: doc/cephfs/mds-journaling.rst [hunk 10]
```diff
@@ -141,14 +141,12 @@ The targetted size of a log segment in terms of number of events is controlled b
 
 .. confval:: mds_log_events_per_segment
 
-The frequency of major segments (noted by the journaling of the latest ``ESubtreeMap``) is controlled by:
+The number of minor mds log segments since last major segment is controlled by:
```
Comment (anthonyeleven) [inline doc/cephfs/mds-journaling.rst:144]: MDS log segments since the last major

File: src/common/options/mds.yaml.in [hunk 11]
```diff
@@ -1741,3 +1731,12 @@ options:
   - mds
   flags:
   - runtime
+- name: mds_log_minor_segments_per_major_segment
+  type: uint
+  level: advanced
+  desc: number of minor segments per major segment.
+  long_desc: The number of minor mds log segments since last major segment after which a major segment is started/logged.
```
Comment (anthonyeleven) [inline src/common/options/mds.yaml.in:1738]: MDS log segments since the last

**Review by anthonyeleven — APPROVED** [review]
Couple of docs requests

**Comment by batrick** [comment]
jenkins test api

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/68744](https://tracker.ceph.com/issues/68744).

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/68786](https://tracker.ceph.com/issues/68786).

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/68859](https://tracker.ceph.com/issues/68859).

**Comment by batrick** [comment]
https://tracker.ceph.com/issues/68859#note-2

**Comment by batrick** [comment]
followup qa fix: #60720

**Pull request merged**